### PR TITLE
New: Concept exercise pacman-rules

### PR DESCRIPTION
**Pull request by pwadsworth**
- Added exemplar, stub and tests.
- Pending: instructions.md (current file is a placeholder)

### CODE REVIEW

**Review by kytrinyx — COMMENTED** [review]
From an Exercism perspective this looks good to me.

@junedev and @angelikatyborska you both have a lot of experience crafting concept exercises. Would you be able to look at this and see if anything jumps out at you with respect to the non-code parts of this?

File: exercises/concept/pacman-rules/.docs/hints.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+## General
+
+- There are three boolean operators to work with boolean values: `not`, `&&`, and `||`.
+- Multiple operators can be combined in a single expression.
+
+## 1. Check if a fast attack can be made
+
+- The boolean operators can be applied to boolean parameters.
+
```
Comment (ErikSchierboom) [inline exercises/concept/pacman-rules/.docs/hints.md:11]:
```suggestion
```

**Review by ErikSchierboom — COMMENTED** [review]
Great! I'll leave the Haskell reviewing part to the Haskell experts. I've just commented on the more generic, concept-exercise related functionality.

File: exercises/concept/pacman-rules/.docs/hints.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# Hints
+
+## General
+
+- There are three boolean operators to work with boolean values: `not`, `&&`, and `||`.
+- Multiple operators can be combined in a single expression.
+
+## 1. Check if a fast attack can be made
```
Comment (angelikatyborska) [inline exercises/concept/pacman-rules/.docs/hints.md:8]: This heading seems to be unrelated to the exercise (sounds to me like it's from  "Annalyn's Infiltration")

**Comment by angelikatyborska** [comment]
I would need to see the instructions to judge 👀 they're the second most important thing after tests (but tests are Haskell code, so I can't judge).

Are the hints done? I only see hints for a single step. In Elixir, if the same hint applies to multiple steps, we just repeat it for each step where it applies. Alternatively they can be put to general hints, but that doesn't show up on the platform when asking for hints for a specific step.

File: exercises/concept/pacman-rules/.docs/introduction.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Introduction
+
+Booleans in Haskell are represented by the `Bool` type, which values can be either `True` or `False`.
+
+Haskell supports three boolean operators: `not` (NOT), `&&` (AND), and `||` (OR).
+
+The three boolean operators each have a different operator precedence. They are evaluated in this order: `not` first, `&&` second, and finally `||`. If you want to override these rules, you can enclose a boolean expression in parentheses (`()`), as the parentheses have an even higher operator precedence.
```
[the remark this answers is not in the document]
Comment (pwadsworth) [reply exercises/concept/pacman-rules/.docs/introduction.md:7]: Good point. Thanks.

File: exercises/concept/pacman-rules/test/Tests.hs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      (loses False True) `shouldBe` True
+
+    it "is False if touching ghost while powered" $ do
+      (loses True True) `shouldBe` False
+
+    it "is False if not touching a ghost" $ do
+      (loses True False) `shouldBe` False
+
+  describe "wins" $ do
+    it "if all dots are eaten" $ do
```
[the remark this answers is not in the document]
Comment (pwadsworth) [reply exercises/concept/pacman-rules/test/Tests.hs:37]: It did originally, I opted to change it because the test output flows better this way 
e.g. 
```
wins
    if all dots are eaten
```
instead of 
```
wins
   is True if all dots are eaten
```
I can change it back if the consistency in the tests code is preferred.

With the False cases I didn't see a way to make it less verbose.

File: exercises/concept/pacman-rules/.docs/hints.md (hunk 2 above)
Comment (pwadsworth) [reply exercises/concept/pacman-rules/.docs/hints.md]: That's right. Edited so it applies to Pacman.

**Comment by pwadsworth** [comment]
> I would need to see the instructions to judge eyes they're the second most important thing after tests (but tests are Haskell code, so I can't judge).
> 
> Are the hints done? I only see hints for a single step. In Elixir, if the same hint applies to multiple steps, we just repeat it for each step where it applies. Alternatively they can be put to general hints, but that doesn't show up on the platform when asking for hints for a specific step.

@angelikatyborska I edited the hints to mirror Elixir's version.

**Comment by pwadsworth** [comment]
@ErikSchierboom all files are complete. Pending decision on how to word the test cases, I think we can merge this and PR #1086 to close #1090.
... unless, of course, I left typos somewhere.

File: exercises/concept/pacman-rules/.docs/hints.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+## 3. Define if pac-man loses
+
+- The function must return a `Boolean` value.
+- You can use the boolean operators `&&` and `not` to combine the arguments for a result.
+
+## 4. Define if pac-man wins
+
+- The function must return a `Boolean` value.
+- You can use the boolean operators `&&` and `not` to combine the arguments and results of one of the previously implemented functions.
```
Comment (MatthijsBlom) [inline exercises/concept/pacman-rules/.docs/hints.md:26]:
```suggestion
- You can use the boolean operators `&&` and `not` to combine the arguments and the result of one of the previously implemented functions.
```

File: exercises/concept/pacman-rules/.docs/instructions.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+In this exercise, you need to translate some rules from the classic game Pac-Man into Haskell functions.
+
+You have four rules to translate, all related to the game states.
+
+> Don't worry about how the arguments are derived, just focus on combining the arguments to return the intended result.
+
+## 1. Define if Pac-Man eats a ghost
+
+Define the `eatsGhost` function that takes two arguments (_if Pac-Man has a power pellet active_ and _if Pac-Man is touching a ghost_) and returns a `Boolean` value if Pac-Man is able to eat the ghost. The function should return `True` only if Pac-Man has a power pellet active and is touching a ghost.
```
Comment (MatthijsBlom) [inline exercises/concept/pacman-rules/.docs/instructions.md:11]:
```suggestion
Define the `eatsGhost` function that takes two arguments (_whether Pac-Man has a power pellet active_ and _whether Pac-Man is touching a ghost_) and returns a `Boolean` value indicating whether Pac-Man is able to eat the ghost. The function should return `True` only when Pac-Man has a power pellet active and is touching a ghost.
```
This may be splitting hairs. Not sure whether it really is clearer to people other than me.

File: exercises/concept/pacman-rules/.meta/config.json [hunk 7]
```diff
@@ -0,0 +1,21 @@
+{
+    "authors": [
+        "pwadsworth",
+        "MatthijsBlom"
```
Comment (MatthijsBlom) [inline exercises/concept/pacman-rules/.meta/config.json:4]: This hardly happened in the end.

File: exercises/concept/pacman-rules/src/PacmanRules.hs [hunk 8]
```diff
@@ -0,0 +1,17 @@
+module PacmanRules (eatsGhost, scores, loses, wins) where
+
+eatsGhost :: Bool -> Bool -> Bool
+eatsGhost =
```
Comment (MatthijsBlom) [inline exercises/concept/pacman-rules/src/PacmanRules.hs:4]: The Elixir original includes parameter names. I am in favor of including them here as well. After all, it is an introductory exercise about _booleans_, not _function declarations_.

File: exercises/concept/pacman-rules/test/Tests.hs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  describe "eatsGhost" $ do
+    it "ghost eaten when touched while powered" $ do
+      (eatsGhost True True) `shouldBe` True
+
+    it "ghost not eaten when not touched" $ do
+      (eatsGhost True False) `shouldBe` False
+
+    it "ghost not eaten when touched un-powered" $ do
+      (eatsGhost False True) `shouldBe` False
+
```
Comment (MatthijsBlom) [inline exercises/concept/pacman-rules/test/Tests.hs:15]: The tests are not exhaustive, but could be. I'm not sure about the likelihood of false positives, but in any case exhaustivity seems cheap.

File: exercises/concept/pacman-rules/.docs/instructions.md (hunk 6 above)
Comment (pwadsworth) [reply exercises/concept/pacman-rules/.docs/instructions.md:11]: If we make the change for one function it should be the same on the following three.

File: exercises/concept/pacman-rules/test/Tests.hs (hunk 9 above)
Comment (pwadsworth) [reply exercises/concept/pacman-rules/test/Tests.hs:15]: I added the rest of the cases.

**Review by angelikatyborska — APPROVED** [review]

**Review by ErikSchierboom — APPROVED** [review]
LGTM The suggestion made by @MatthijsBlom might be worth applying.

@MatthijsBlom You happy with the Haskell part of the PR?

File: exercises/concept/pacman-rules/.docs/instructions.md (hunk 6 above)
Comment (MatthijsBlom) [reply exercises/concept/pacman-rules/.docs/instructions.md:11]: Yes. I have a patch ready; if it is decided that _whether_ & _when_ are prefered I can push it.

@ErikSchierboom Do I understand correctly that you meant this suggestion?

**Comment by MatthijsBlom** [comment]
One other thing: going by [these docs](https://exercism.org/docs/building/tracks/concept-exercises) and what @ErikSchierboom said about it on Slack, it seems we should have a `.docs/introduction.md.tpl` rather than a `.docs/introduction.md`.

Everything I did not comment on carries my approval, yes.

**Comment by ErikSchierboom** [comment]
> One other thing: going by [these docs](https://exercism.org/docs/building/tracks/concept-exercises) and what @ErikSchierboom said about it on Slack, it seems we should have a .docs/introduction.md.tpl rather than a .docs/introduction.md.

We could definitely do that. It's also fine to do this as a follow-up PR (which could also update the other two concept exercises).

> Everything I did not comment on carries my approval, yes.

Thanks!

@pwadsworth This PR has three merge conflicts. Once those are fixed, we can merge!

**Comment by pwadsworth** [comment]
@ErikSchierboom I applied @MatthijsBlom suggestion throughout `instructions.md` and resolved the conflicts.

**Comment by kytrinyx** [comment]
@pwadsworth It looks like there are some linting errors on `exercises/concept/pacman-rules/test/Tests.hs`. Would you take a look?

**Comment by pwadsworth** [comment]
@kytrinyx I removed the offending parenthesis. For future reference, should I go by the linter warnings over a more clear (though redundant) style in the Test files?

**Comment by kytrinyx** [comment]
@pwadsworth Yes, the checks will fail if HLint fails.

**Comment by petertseng** [comment]
When is it planned to merge this branch into main?

**Comment by petertseng** [comment]
EBWODP: Rephrased question to be clearer

What is the future plan of this pacman-rules branch? When will the pacman-rules branch be merged into main?

**Comment by kytrinyx** [comment]
I will rebase and squash and open a PR.
